File: algorithms/warmup/fib_part_sum_ld.py
#python3                                                                      
#where m ≤ n, find the last digit of the sum F m + F m+1 + Fn

def fib_mod_sum(m, n):                                                           
    mod_seq = [0, 1]                                                          
    # mod is 10 -> pisano period is 60                                        
    pp = 60                                                                   
                                                                              
    fib_m = m % pp
    fib_n = n % pp
                                                                              
    for _ in range(1, fib_n):                                                 
        mod_seq.append((mod_seq[-1] + mod_seq[-2]) % 10)                      
                                                                              
    return sum(mod_seq[fib_m:fib_n + 1]) % 10                                      
                                                                              
                                                                              
def main():                                                                   
    m, n = map(int, input().split())                                                          
    print(fib_mod_sum(m, n))                                                     
                                                                              
                                                                              
if __name__ == "__main__":                                                    
    main()                                                                    

File: algorithms/warmup/fib_part_sum_ld.cpp
#include <iostream>
#include <vector>
#include <numeric>


int fib_mod(int64_t &m, int64_t &n) {
    // the length of pisano_period(10) is 60
    int pp = 60;
    int fib_m = m % pp;
    int fib_n = n % pp;

    std::vector<int> mod_seq = {0, 1};

    for (int i = 1; i < fib_n; i++) {
        int size = mod_seq.size();
        int res = (mod_seq.back() + mod_seq[size - 2]) % 10;
        mod_seq.push_back(res);
    }
    int sum = 0;

    for (int i = fib_m; i <= fib_n; i++) {
        sum += mod_seq[i];
    }

    return sum % 10;
}

int main() {
    int64_t m, n;
    std::cin >> m >> n;
    std::cout << fib_mod(m, n);

    return 0;
}
